cpp/examples/tpch/q1.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/ast/expressions.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/scalar/scalar.hpp>
+
+#include "utils.hpp"
+
+/*
+create view lineitem as select * from '~/tpch_sf1/lineitem/part-0.parquet';
+
+select
+    l_returnflag,
+    l_linestatus,
+    sum(l_quantity) as sum_qty,
+    sum(l_extendedprice) as sum_base_price,
+    sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,
+    sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge,
+    avg(l_quantity) as avg_qty,
+    avg(l_extendedprice) as avg_price,
+    avg(l_discount) as avg_disc,
+    count(*) as count_order
+from
+    lineitem
+where
+        l_shipdate <= date '1998-09-02'
+group by
+    l_returnflag,
+    l_linestatus
+order by
+    l_returnflag,
+    l_linestatus;
+*/
+
+std::unique_ptr<cudf::column> calc_disc_price(std::unique_ptr<table_with_cols>& table) {
```

Comment:
A few notes:
- Sprinkle in `const` everywhere it is appropriate to do so, in the parameters and function body. It makes reasoning about the query easier. This applies throughout.
- How much "magic" is acceptable here, considering what metadata could come from a query engine? For example, can we _compute_ the scale of `-4` for the discount price decimal type instead of using a raw value `-4`?
- Should these functions accept and pass through `stream` / `mr` parameters? (It's okay if the answer is no, so long as we're able to benchmark what we want while using default memory resources and default streams.)

Reply:
Now determining scale using `cudf::binary_operation_fixed_point_scale`

Reply:
Echoing @bdice 's comments, all functions should take a stream and a memory resource as input arguments and the resulting `std::unique_ptr<cudf::column>` should be constructed from the given mr.

for your reference:

- https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#streams-streams
- https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#output-memory

Reply:
For the `mr` argument, I am setting global memory resource now using 
```
    auto resource = create_memory_resource(args.memory_resource_type);
    rmm::mr::set_current_device_resource(resource.get());
```

Does that work for the memory resources? I would look into adding streams to the functions, that is something we would probably need going ahead.

Reply:
Added `stream` / `mr` to wherever necessary